### DISCUSSION TITLE
Apply hook/middleware to requests with wildcard or path parameters

### DISCUSSION THREAD

**Discussion by SimenB** [fastify/fastify#2939] [Q&A]
Hiya!

I'm currently trying to migrate an Express app to Fastify, and was wondering if there is some way to use URL/path matching in middlewares?

Code we have today in express:

```js
// Assets used for e-mails
app.get(['/assets/fonts/*', '/assets/mail/*'], (_, res, next) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.removeHeader('Cross-Origin-Resource-Policy');
  next();
});
```

(ignore array thing, that's not what I'm having issues with, except for it resulting in `TypeError`s 😅).

So essentially, for any requests that match the patterns provided, we wanna set a header and remove another. In fastify, there is no `next` passed to handlers, so I cannot just use that. But if I rewrite this to a `onRequest` hook, I have to manually check both http method, and attempt to match on `request.url`.
I cannot really use a handler for this as our handler is just `fastify.next('/*')` so we don't explicitly handle any of the requests ourselves.

Is there any way of either registering a hook for paths with wildcards (e.g. `prefix` passed to `register`) or potentially some util from fastify/find-my-way which is essentially `fastify.matchUrl('/assets/{fonts,mail}/*', req.url)` or some such?

**Comment by mcollina** [comment]
You could do something with the `onRoute`  flag and https://github.com/fastify/fastify-static#wildcard. Setting `wildcard: false` will enable you to receive an `onRoute` call for each file/route. Then, you can just add that logic as a `onRequest` handler there.

**Reply by SimenB** [reply]
Then I'd have to serve those files "manually", right? ATM we leave that to Next (I believe they do some hashing etc, so it's not a path we necessarily know at design time)

**Reply by mcollina** [reply]
> Then I'd have to serve those files "manually", right? ATM we leave that to Next (I believe they do some hashing etc, so it's not a path we necessarily know at design time)

Yes exactly.

Unfortunately this leaves it for you to do the regexp checking yourself for each incoming request, which is what Express does anyway.

```js
app.addHook('onRequest', async function (req, reply) {
  if (reply.indexOf('/assets/fonts/') === 0 || reply.indexOf('/assets/fonts/') === 0) {
      reply.header('Access-Control-Allow-Origin', '*')
  }
})
```

I'm not sure it's the right place for doing ` res.removeHeader('Cross-Origin-Resource-Policy')`.

**Reply by SimenB** [reply]
Yep, I was sorta expecting having to write something like that. Seems sensible enough 👍 Thanks!

Thoughts on fastify exporting or otherwise exposing some sort of `matchPath` thing? I tried to look at `find-my-way` but from what I could see it was mostly "find handler for this request" without some intermediary "match req.url with path" exposed

**Reply by mcollina** [reply]
I would hardly recommend against it. It's one of the main reason that causes express applications to slow down.
